File: server/app/signaling.py
import json

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from .config import settings
from .rooms import Peer, RoomManager

manager = RoomManager(settings.room_ttl_seconds, settings.max_room_size)

router = APIRouter()


@router.post("/rooms")
def create_room() -> dict:
    room = manager.create()
    return {"room_id": room.id, "link": f"/?room={room.id}"}


@router.websocket("/ws/{room_id}")
async def signaling(websocket: WebSocket, room_id: str, peer_id: str) -> None:
    room = manager.get(room_id)
    if room is None:
        await websocket.close(code=4404)
        return

    await websocket.accept()
    peer = Peer(id=peer_id, websocket=websocket)
    if not manager.add_peer(room, peer):
        await websocket.send_json({"type": "error", "message": "room full"})
        await websocket.close()
        return

    try:
        existing = [p.id for p in manager.peers(room, exclude=peer_id)]
        await websocket.send_json({"type": "welcome", "peer_id": peer_id, "peers": existing})

        for p in manager.peers(room, exclude=peer_id):
            await p.websocket.send_json({"type": "peer-joined", "peer_id": peer_id})

        while True:
            raw = await websocket.receive_text()
            message = json.loads(raw)
            mtype = message.get("type")

            if mtype == "signal":
                target = room.peers.get(message.get("to"))
                if target is not None:
                    await target.websocket.send_json(
                        {
                            "type": "signal",
                            "from": peer_id,
                            "data": message.get("data"),
                        }
                    )
            elif mtype == "leave":
                break

    except WebSocketDisconnect:
        pass
    finally:
        manager.remove_peer(room, peer_id)
        for p in manager.peers(room):
            await p.websocket.send_json({"type": "peer-left", "peer_id": peer_id})

File: apps/mobile/lib/services/signaling.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:web_socket_channel/web_socket_channel.dart';

/// Payload relayed between peers via the signaling server's `signal` messages.
class ChatPayload {
  const ChatPayload({required this.text});

  final String text;

  Map<String, dynamic> toJson() => {'kind': 'chat', 'text': text};

  static ChatPayload? tryFrom(Object? data) {
    if (data is! Map<String, dynamic>) return null;
    if (data['kind'] != 'chat' || data['text'] is! String) return null;
    return ChatPayload(text: data['text'] as String);
  }
}

sealed class SignalEvent {
  const SignalEvent();
}

class WelcomeEvent extends SignalEvent {
  const WelcomeEvent({required this.peers});

  final List<String> peers;
}

class PeerJoinedEvent extends SignalEvent {
  const PeerJoinedEvent({required this.peerId});

  final String peerId;
}

class PeerLeftEvent extends SignalEvent {
  const PeerLeftEvent({required this.peerId});

  final String peerId;
}

class SignalRelayEvent extends SignalEvent {
  const SignalRelayEvent({required this.from, required this.data});

  final String from;
  final Object? data;
}

class SignalClosedEvent extends SignalEvent {
  const SignalClosedEvent({required this.code, required this.reason});

  final int code;
  final String reason;
}

/// The parts of [Signaling] that a chat screen depends on, so it can be
/// substituted for unit/widget testing without a live network connection.
abstract interface class SignalClient {
  Stream<SignalEvent> get events;

  /// Relays [data] to peer [to] via the signaling server.
  void send(String to, Object data);

  void close();
}

class Signaling implements SignalClient {
  Signaling._(
    this.backendUrl,
    this.roomId,
    this.peerId,
    this._channel,
    this._events,
  );

  final String backendUrl;
  final String roomId;
  final String peerId;
  final WebSocketChannel _channel;
  final StreamController<SignalEvent> _events;

  @override
  Stream<SignalEvent> get events => _events.stream;

  /// Creates a room on the signaling server and returns its id.
  static Future<String> createRoom(String backendUrl) async {
    final res = await http
        .post(Uri.parse('$backendUrl/rooms'))
        .timeout(const Duration(seconds: 10));
    if (res.statusCode != 200) {
      throw Exception('create room failed: HTTP ${res.statusCode}');
    }
    final body = jsonDecode(res.body) as Map<String, dynamic>;
    return body['room_id'] as String;
  }

  static Uri _websocketUri(String backendUrl, String roomId, String peerId) {
    final base = Uri.parse(backendUrl);
    final scheme = base.scheme == 'https' ? 'wss' : 'ws';
    return Uri(
      scheme: scheme,
      host: base.host,
      port: base.port,
      pathSegments: ['ws', roomId],
      queryParameters: {'peer_id': peerId},
    );
  }

  static Signaling connect({
    required String backendUrl,
    required String roomId,
    required String peerId,
  }) {
    final events = StreamController<SignalEvent>();
    final channel = WebSocketChannel.connect(
      _websocketUri(backendUrl, roomId, peerId),
    );
    final signaling = Signaling._(backendUrl, roomId, peerId, channel, events);

    channel.stream.listen(
      (raw) {
        try {
          final message = jsonDecode(raw as String) as Map<String, dynamic>;
          final type = message['type'];
          switch (type) {
            case 'welcome':
              events.add(WelcomeEvent(
                peers: List<String>.from(message['peers'] as List),
              ));
            case 'peer-joined':
              events.add(PeerJoinedEvent(peerId: message['peer_id'] as String));
            case 'peer-left':
              events.add(PeerLeftEvent(peerId: message['peer_id'] as String));
            case 'signal':
              events.add(SignalRelayEvent(
                from: message['from'] as String,
                data: message['data'],
              ));
            default:
              break;
          }
        } catch (_) {
          // ignore malformed frames
        }
      },
      onError: (_) => events.add(const SignalClosedEvent(code: 1006, reason: '')),
      onDone: () => events.add(const SignalClosedEvent(code: 1000, reason: '')),
    );

    return signaling;
  }

  /// Relays [data] to peer [to] via the signaling server.
  @override
  void send(String to, Object data) {
    _channel.sink.add(jsonEncode({'type': 'signal', 'to': to, 'data': data}));
  }

  @override
  void close() {
    _channel.sink.add(jsonEncode({'type': 'leave'}));
    _channel.sink.close();
  }
}